### PR TITLE
Allow Changing of Player Sample in ServerListPingEvent

### PR DESCRIPTION
**Pull request by willies952002**
This will allow plugins to change what a player sees when they hover over the player count on the multiplayer server list when the `ServerListPingEvent` is fired.

### CODE REVIEW

**Comment by 0-x-2-2** [comment]
This would be fantastic.

**Comment by zachbr** [comment]
I would rather see this just accept player objects until we get our own profile API in place.
We can add some to-do comments in place to handle profiles until then.

**Comment by willies952002** [comment]
I don't see how using player objects would be practical in the sense that this is meant for adding the ability to easily modify the "Player Sample", which various servers have taken advantage of by using either Packets or Bungeecord's ability to modify it directly during the `ProxyPingEvent`, allowing them to do things such as the following example:
this

If anything, I could switch it to accepting simple Strings then build the profiles from there in server code

**Comment by kashike** [comment]
I don't agree with using `Player`s.

**Comment by zachbr** [comment]
I don't really like it either but this shouldn't be both a ServerListPingEvent PR and an authlib profile API PR.

Edit: open to alternatives.

**Comment by electronicboy** [comment]
Wouldn't it make more sense to expose that as a Map<String, Player>, Player included so that we at least know who is inside that, as well as a String so that they can actually control what is going to be presented to the client. So long as the code following the event accepts that the value is potentially going to be null, I feel that that is a sane option and would likely be much more preferred by developers (and more in line with existing API design/behavior)

nb: Implementation wise, we would specifically need to use a LinkedHashMap, to retain insertion order, not entirely the prettiest option however this implementation is somewhat standard for Bukkit... Could have some wrapper object in an arraylist exposed by the event, that would likely permit for a more "in line with what you're expecting", as well as make it easier for handling order without pretty much requiring a new map to be created for a trivial change....

File: Spigot-API-Patches/0051-Allow-Changing-of-Player-Sample-in-ServerListPingEve.patch [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ 
+     public ServerListPingEvent(final InetAddress address, final String motd, final int numPlayers, final int maxPlayers) {
+         super(); // Paper - Is this event being fired async?
+@@ -145,4 +146,13 @@ public class ServerListPingEvent extends ServerEvent implements Iterable<Player>
+     public Iterator<Player> iterator() throws UnsupportedOperationException {
+         throw new UnsupportedOperationException();
+     }
++
++    // Paper start
++    public void setSample(java.util.List<com.mojang.authlib.GameProfile> profiles) {
```
Comment (aikar) [inline Spigot-API-Patches/0051-Allow-Changing-of-Player-Sample-in-ServerListPingEve.patch]: See comment feedback

**Review by aikar — CHANGES_REQUESTED** [review]
I think this API for now and the future only needs to be a `List<String>`, and maybe named `setSampleText(List<String>);`

Since the profiles are not bound to real players, there's no benefit to using Profiles or Player objects.

Lets just use Strings, which removes the need to add the authlib to the PR.

**Review by aikar — APPROVED** [review]

**Comment by aikar** [comment]
Manually merged in c86f61f8afd3b99d490e11690091aca5731d28b0

**Comment by RoboMWM** [comment]
So this broke plugins that iterate over the event to remove players from the list (and thus the player count that appears). Are all plugins supposed to do what EssentialsX does (detect if paper, and use reflection to get and remove players) https://github.com/drtshock/Essentials/blob/2.x/Essentials/src/com/earth2me/essentials/EssentialsServerListener.java or should compatibility be added?

**Comment by zachbr** [comment]
Actually no. I don't want this on this PR.

**Pull request closed without merging**